### PR TITLE
PSP-6494 : Acquisition List View Export (Excel)

### PR DESCRIPTION
**Pull request by eddherrera**
image

image

### CODE REVIEW

File: source/backend/api/Areas/Acquisition/Controllers/ReportController.cs
Comment (asanchezr) [inline source/backend/api/Areas/Acquisition/Controllers/ReportController.cs:32]: @eddherrera - we have a separate folder for "reports" - we put them under `api/Areas/Reports/Controllers/AcquisitionController.cs`. Please consider moving it there (or merging with the one controller from Devin's PR once that gets merged

Comment (eddherrera) [reply source/backend/api/Areas/Acquisition/Controllers/ReportController.cs:32]: Ok, I'll merge after his.

Comment (eddherrera) [reply source/backend/api/Areas/Acquisition/Controllers/ReportController.cs]: relocated

**Comment by devinleighsmith** [comment]
@eddherrera filename does not match AC
actual: pims-acquisition-files.xlsx
expected: Acquisition_File_Export.xlsx

**Comment by devinleighsmith** [comment]
AlejandroSanchez, BobBillySmith

**Comment by devinleighsmith** [comment]
Worksheet name does not match AC:
actual: PIMS_Acquisition_Files
expected: Acquisition File Export

**Comment by devinleighsmith** [comment]
acquisition file number appears to be the id of the acquisition file, which is incorrect. Should be the file number aka: 
image

File: source/backend/api/Helpers/Middleware/ErrorHandlingMiddleware.cs
Comment (devinleighsmith) [inline source/backend/api/Helpers/Middleware/ErrorHandlingMiddleware.cs:184]: would a 404 make more sense? that is what we throw when a user requests an id that does not exist.

File: source/backend/dal/Helpers/Extensions/AddressExtensions.cs
Comment (devinleighsmith) [inline source/backend/dal/Helpers/Extensions/AddressExtensions.cs:46]: why no if statement check here? is this a required field in the DB?

File: source/backend/dal/Helpers/Extensions/PersonExtensions.cs
Comment (devinleighsmith) [inline source/backend/dal/Helpers/Extensions/PersonExtensions.cs:16]: see my other comment, seems to be missing spaces.

File: source/backend/dal/Models/AcquisitionFileExportDto.cs
Comment (devinleighsmith) [inline source/backend/dal/Models/AcquisitionFileExportDto.cs:34]: nit (spacing)

File: source/backend/dal/Repositories/AcquisitionFileRepository.cs
Comment (devinleighsmith) [inline source/backend/dal/Repositories/AcquisitionFileRepository.cs:74]: This seems dangerous, as we now have two filters, one that returns data for the acquisition file list view, and this one that returns data for the export. there is no guarantee that the two will return the same data, or remain in sync in the case of future updates. 

Can you explain why we need this function instead of just re-using the existing acquisition search that populates the list view? even if we don't reuse that entire repository method could we at least extract out the database query part to make sure they are the same?

File: source/frontend/src/features/acquisition/hooks/useAcquisitionFileExport.ts
Comment (devinleighsmith) [inline source/frontend/src/features/acquisition/hooks/useAcquisitionFileExport.ts:37]: console.log

File: source/frontend/src/features/acquisition/hooks/useAcquisitionFileExport.ts
Comment (devinleighsmith) [inline source/frontend/src/features/acquisition/hooks/useAcquisitionFileExport.ts:22]: I don't think that csv is supported here so can probably simplify this.

File: source/frontend/src/features/acquisition/hooks/useAcquisitionFileExport.ts
Comment (devinleighsmith) [inline source/frontend/src/features/acquisition/hooks/useAcquisitionFileExport.ts:42]: ?

File: source/frontend/src/hooks/pims-api/useApiAcquisitionFile.ts
Comment (devinleighsmith) [inline source/frontend/src/hooks/pims-api/useApiAcquisitionFile.ts:47]: same comment here regarding csv, pretty sure this will throw an error from the backend if you request csv.

File: source/backend/dal/Helpers/Extensions/PersonExtensions.cs
Comment (devinleighsmith) [inline source/backend/dal/Helpers/Extensions/PersonExtensions.cs:8]: is this necessary give that GetFullName extension method already exists?

File: source/backend/dal/Helpers/Extensions/AcquisitionOwnerExtensions.cs
Comment (devinleighsmith) [inline source/backend/dal/Helpers/Extensions/AcquisitionOwnerExtensions.cs:10]: Personally I prefer this syntax for conciseness/readibility, what do you think?

string[] names = { person.FirstName, person.MiddleNames, person.Surname };
return string.Join(" ", names.Where(n => n != null && n.Trim() != string.Empty));

**Review by devinleighsmith — CHANGES_REQUESTED** [review]
I'm mostly concerned about not re-using the acquisition list view filter method, the two filters should return the same results.

File: source/frontend/src/features/acquisition/hooks/useAcquisitionFileExport.ts
Comment (eddherrera) [reply source/frontend/src/features/acquisition/hooks/useAcquisitionFileExport.ts:37]: removed

File: source/frontend/src/features/acquisition/hooks/useAcquisitionFileExport.ts
Comment (eddherrera) [reply source/frontend/src/features/acquisition/hooks/useAcquisitionFileExport.ts:22]: simplified,

File: source/frontend/src/features/acquisition/hooks/useAcquisitionFileExport.ts
Comment (eddherrera) [reply source/frontend/src/features/acquisition/hooks/useAcquisitionFileExport.ts:42]: fixed

File: source/backend/dal/Helpers/Extensions/PersonExtensions.cs
Comment (eddherrera) [reply source/backend/dal/Helpers/Extensions/PersonExtensions.cs:8]: the existing extension method doesn't include Suffix

**Comment by eddherrera** [comment]
> I'm mostly concerned about not re-using the acquisition list view filter method, the two filters should return the same results.

Updated with a new private static method in the repository. also other comments addressed.

File: source/backend/dal/Repositories/AcquisitionFileRepository.cs
Comment (devinleighsmith) [inline source/backend/dal/Repositories/AcquisitionFileRepository.cs:116]: I don't know what others think @asanchezr or @FuriousLlama but I can't get behind this syntax. This also seems to be conflating getting data into entities with transforming entities into a display format all within a repository which seems like a violation of separation of concerns.
I would prefer to see:
1) controller handles formatting models, validating requests and returning http responses
2) service handles composing repo calls and responses, business logic, and authorization.
3) repo handles getting data from database or external services in whatever format that db/service returns that data in.

File: source/backend/dal/Repositories/AcquisitionFileRepository.cs
Comment (asanchezr) [reply source/backend/dal/Repositories/AcquisitionFileRepository.cs:83]: I agree that we should try to use the existing "GetPage" functionality of acquisition list view. If we wanted to get all records instead of pages we could send `PageSize = Int.MaxValue` (or whatever C# constant is for max integer). 

I don't see the need for a separate function and as Devin mentions I would be concerned about the two functions not returning the same data

File: source/backend/dal/Repositories/AcquisitionFileRepository.cs
Comment (asanchezr) [reply source/backend/dal/Repositories/AcquisitionFileRepository.cs:116]: Agreed. I would prefer that the Entity -> DTO conversion is done at the Controller level. The repo should always return DB entities to keep separation of concerns. 
Also a minor note on the export DTO name - we usually do not use "DTO" in their names (even if they are DTOs in purpose). A more fitting name would be **AcquisitionFileExportModel**

Comment (asanchezr) [reply source/backend/dal/Repositories/AcquisitionFileRepository.cs:116]: Forgot to mention that I don't think it would be necessary to create a Mapster mapping for this export model conversion. If you look at Devin's PR you can pass the entity in the constructor and create an export model from it. I am using that pattern as well in my export H120 story. 

It would make sense to use mappings for things that are meant to be shared/composed. These "export" models are one-offs so no need to create mappings for them (in my opinion)

File: source/frontend/src/features/acquisition/hooks/useAcquisitionFileExport.ts
Comment (asanchezr) [inline source/frontend/src/features/acquisition/hooks/useAcquisitionFileExport.ts:45]: A couple issues I see here:
- I would question why are we triggering the download in the "use..." hook. I'd rather just return the raw blob data and let the consumer of the hook decide what to do with it. Please see Devin's PR for an example. I'm doing the same in my export PR. This hook should be concerned with calling the API and returning the response - not with also triggering the file save popup.
- I see this on line 42 `catchAxiosError(e, dispatch, actionTypes.DELETE_PARCEL);` - is this copy/paste? This hook is not related to parcels at all...

Comment (asanchezr) [reply source/frontend/src/features/acquisition/hooks/useAcquisitionFileExport.ts:45]: One more comment. I also don't think this new functionality warrants a separate "use..." hook. The method **exportAcquisitionFiles** could be added to the **useAcquisitionProvide** hook. The functionality is sufficiently relevant to that provider that it should go there instead of a separate hook with just one function

File: source/backend/api/Helpers/Middleware/ErrorHandlingMiddleware.cs
Comment (asanchezr) [reply source/backend/api/Helpers/Middleware/ErrorHandlingMiddleware.cs:184]: I also question the need for another exception. To me 404 would be sufficient. Or another option I am pondering on my PR is to use a non-error status code **204 No Content** - because in theory the API shouldn't be concerned that the frontend wants to display a warning if no results are found. 
In my opinion, the API should not return an error code because no error happened, it just didn't find any matching records. 

What do you think of using 204 instead of 400/404 here @eddherrera and @devinleighsmith ?

File: source/frontend/src/features/acquisition/hooks/useAcquisitionFileExport.ts
Comment (eddherrera) [reply source/frontend/src/features/acquisition/hooks/useAcquisitionFileExport.ts:45]: This is the patter used for exporting leases. Not the first time I get comments for using patterns used in other sections of the code.

File: source/backend/dal/Repositories/AcquisitionFileRepository.cs
Comment (eddherrera) [reply source/backend/dal/Repositories/AcquisitionFileRepository.cs:116]: Should I move the query to the properties repository since the query is returning one row per property in the Acquisition file?

File: source/frontend/src/features/acquisition/hooks/useAcquisitionFileExport.ts
Comment (asanchezr) [reply source/frontend/src/features/acquisition/hooks/useAcquisitionFileExport.ts:45]: Sorry if my comments came in as a bit harsh - totally not my intention. As I see it, PR comments are more "suggestions" than anything (unless it is some glaring error in the code). In my view, separating the download from the useXXX hook makes it more re-usable. Also, leases is the oldest portion of the codebase so the patterns used there might have been refined lately

File: source/backend/api/Areas/Reports/Controllers/AcquisitionController.cs
Comment (asanchezr) [inline source/backend/api/Areas/Reports/Controllers/AcquisitionController.cs:117]: nit: not sure if we would need to check for acquisitionFileData to not be null?

File: source/backend/api/Services/AcquisitionFileService.cs
Comment (asanchezr) [inline source/backend/api/Services/AcquisitionFileService.cs:93]: nit: I think you need to include a placeholder in the log statement - i.e.

```csharp
_logger.LogDebug("Acquisition file search with filter: {filter}", filter);
```

Comment (eddherrera) [reply source/backend/api/Services/AcquisitionFileService.cs:93]: Agreements doesn't has one...

image

Now that I think about it, shouldn't it be ??

`_logger.LogDebug("Acquisition file search with filter: {0}", filter);  `

File: source/backend/dal/Repositories/AcquisitionFileRepository.cs
Comment (asanchezr) [inline source/backend/dal/Repositories/AcquisitionFileRepository.cs:60]: nit: do we want to call ToList() here? That would bring the whole list into memory (instead of one page of results)...

The old code worked by doing the Skip, Take on the original query, and then calling ToList() - could we do the same here?

File: source/backend/api/Areas/Reports/Controllers/AcquisitionController.cs
Comment (eddherrera) [reply source/backend/api/Areas/Reports/Controllers/AcquisitionController.cs:117]: it doesn't seem so. It just returns an empty list instead of a null.

File: source/backend/api/Services/AcquisitionFileService.cs
Comment (eddherrera) [reply source/backend/api/Services/AcquisitionFileService.cs]: updated

File: source/backend/dal/Repositories/AcquisitionFileRepository.cs
Comment (eddherrera) [reply source/backend/dal/Repositories/AcquisitionFileRepository.cs]: removed

File: source/backend/dal/Repositories/AcquisitionFileRepository.cs
Comment (eddherrera) [reply source/backend/dal/Repositories/AcquisitionFileRepository.cs]: updated

File: source/backend/api/Services/AcquisitionFileService.cs
Comment (asanchezr) [inline source/backend/api/Services/AcquisitionFileService.cs:113]: **nit:** I see mixed usage of `!= null` and `xyz is not null` - would appreciate sticking to one pattern

File: source/backend/api/Services/AcquisitionFileService.cs
Comment (asanchezr) [inline source/backend/api/Services/AcquisitionFileService.cs:122]: I think there's potential for NPE errors by not checking for nulls on a few properties here:

- fileProperty.file.AcquisitionFileStatusTypeCodeNavigation.Description
- fileProperty.file.PimsAcquisitionFilePeople.Select(...)
- fileProperty.file.PimsAcquisitionOwners.Select(...)

File: source/frontend/src/features/acquisition/hooks/useAcquisitionFileExport.ts
Comment (asanchezr) [inline source/frontend/src/features/acquisition/hooks/useAcquisitionFileExport.ts:43]: not sure DELETE_PARCEL action type is relevant here - what action should we use instead?

File: source/frontend/src/features/acquisition/list/AcquisitionListView.tsx
Comment (asanchezr) [inline source/frontend/src/features/acquisition/list/AcquisitionListView.tsx:61]: I think there was a previous comment about removing all references to CSV

File: source/frontend/src/hooks/pims-api/useApiAcquisitionFile.ts
Comment (asanchezr) [inline source/frontend/src/hooks/pims-api/useApiAcquisitionFile.ts:55]: csv format not supported by our export endpoints (currently)

**Review by asanchezr — APPROVED** [review]

File: source/backend/api/Services/AcquisitionFileService.cs
Comment (eddherrera) [reply source/backend/api/Services/AcquisitionFileService.cs]: question raised to Julian. Changes should be handled with new story if needed.

File: source/backend/api/Services/AcquisitionFileService.cs
Comment (eddherrera) [reply source/backend/api/Services/AcquisitionFileService.cs]: updated

File: source/backend/api/Services/AcquisitionFileService.cs
Comment (eddherrera) [reply source/backend/api/Services/AcquisitionFileService.cs]: Added check for status. each item is checked and collection are initialized as empty list from the parent constructor.

File: source/frontend/src/features/acquisition/hooks/useAcquisitionFileExport.ts
Comment (eddherrera) [reply source/frontend/src/features/acquisition/hooks/useAcquisitionFileExport.ts]: changed to log error.

File: source/frontend/src/features/acquisition/list/AcquisitionListView.tsx
Comment (eddherrera) [reply source/frontend/src/features/acquisition/list/AcquisitionListView.tsx]: removed

File: source/frontend/src/hooks/pims-api/useApiAcquisitionFile.ts
Comment (eddherrera) [reply source/frontend/src/hooks/pims-api/useApiAcquisitionFile.ts]: removed

**Review by devinleighsmith — APPROVED** [review]

**Pull request merged**
